src/libraries/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.cs

```diff
@@ -84,13 +71,28 @@ protected override bool ReleaseHandle()
             return true;
         }
 
-        internal void CloseAsIs(bool abortive)
+        internal void CloseAsIs(bool abortive, bool finalizing)
         {
 #if DEBUG
             // If this throws it could be very bad.
             try
             {
 #endif
+                // When the handle was not released due it being used, we try to make those on-going calls return.
+                // TryUnblockSocket will unblock current operations but it doesn't prevent
+                // a new one from starting. So we must call TryUnblockSocket multiple times.
+                //
+                // When the Socket is disposed from the finalizer thread
+                // it is no longer used for operations and we can skip TryUnblockSocket fall back to ReleaseHandle.
+                // This avoids blocking the finalizer thread when TryUnblockSocket is unable to get the reference count to zero.
+                if (finalizing)
```

Comment:
Since all we are doing is calling Dispose and then returning, why do we even call CloseAsIs?

Why don't we just modify Socket.Dispose(bool disposing) to just do:

if (!disposing)
{
  handle.Dispose();
  return;
}

That seems equivalent to this code, and much simpler.